encode.go

```diff
@@ -114,6 +120,9 @@ func (enc *Encoder) encode(key Key, rv reflect.Value) {
 	case time.Time, TextMarshaler:
 		enc.keyEqElement(key, rv)
 		return
+
+	case Marshaler:
```

Comment:
If a user-defined type is a Marshaler and a TextMarshaler, the former should take precedence.